### PR TITLE
Add proper RGBW led support

### PR DESCRIPTION
**Pull request by kwerdenker**
While QMK allowed to use RGBW LEDs for lighting effects, the additional white channel was not actually used.
This commit changes this by adding a converter function to changw RGB colors into RGBW colors and the necessary glue logic for it to be used.

Also some of the color conversion/LED setting functions in the effects logic had to be adjusted to work with the new system.

All changes related to RGBW support are gated by the appropriate RBGW define already in use.

### CODE REVIEW

File: quantum/rgblight.c [hunk 1]
```diff
@@ -118,6 +141,14 @@ void setrgb(uint8_t r, uint8_t g, uint8_t b, LED_TYPE *led1) {
   (*led1).b = b;
 }
 
+#ifdef RGBW
+void setrgbw(uint8_t r, uint8_t g, uint8_t b, uint8_t w, LED_TYPE *led1) {
+  (*led1).r = r;
+  (*led1).g = g;
+  (*led1).b = b;
+  (*led1).w = w;
```
[the remark this answers is not in the document]
Comment (kwerdenker) [reply quantum/rgblight.c]: That function is a modified copy of the pre-existing function `setrgb` , so I kept the styling "as is". Can I assume this change should be applied to that function aswell?

File: quantum/rgblight.c [hunk 2]
```diff
@@ -57,59 +57,82 @@ rgblight_config_t rgblight_config;
...
-void sethsv(uint16_t hue, uint8_t sat, uint8_t val, LED_TYPE *led1) {
-  uint8_t r = 0, g = 0, b = 0, base, color;
+#ifdef RGBW
+void convert_rgb_to_rgbw(uint8_t* r, uint8_t* g, uint8_t* b, uint8_t* w) {
+    // Determine lowest value in all three colors, put that into
+    // the white channel and then shift all colors by that amount
+    *w = MIN(*r, MIN(*g, *b));
+    *r = *r - *w;
+    *g = *g - *w;
+    *b = *b - *w;
```
[the remark this answers is not in the document]
Comment (kwerdenker) [reply quantum/rgblight.c]: I personally prefer the more explicit way because it's easier to read and understand at a glance while the compiler most likely generates the same instructions from both. If the proposed change is the preferred one for the project though, I can apply it.

File: quantum/rgblight.c [hunk 3]
```diff
@@ -527,6 +571,9 @@ void rgblight_set(void) {
       led[i].r = 0;
       led[i].g = 0;
       led[i].b = 0;
+      #ifdef RGBW
```
[the remark this answers is not in the document]
Comment (kwerdenker) [reply quantum/rgblight.c]: This is not consistent in the original source. Compare for example the way it's done in the `rgblight_set` function, where the #ifdef's are placed intended with the code. Shall those instances be moved to the beginning of the line too?

File: quantum/rgblight.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (vomindoraan) [reply quantum/rgblight.c]: Yes, I recommend changing those occurrences as well.

[the remark this answers is not in the document]
Comment (kwerdenker) [reply quantum/rgblight.c]: Has been fixed with 0f457c1

File: quantum/rgblight.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (kwerdenker) [reply quantum/rgblight.c]: Has been fixed with 0f457c1

File: quantum/rgblight.c [hunk 4]
```diff
@@ -57,59 +57,82 @@ rgblight_config_t rgblight_config;
...
+    *b = *b - *w;
+}
+#endif
 
-  if (val > RGBLIGHT_LIMIT_VAL) {
-      val=RGBLIGHT_LIMIT_VAL; // limit the val
-  }
+void convert_hsv_to_rgb(uint16_t hue, uint8_t sat, uint8_t val, uint8_t *r_p, uint8_t *g_p, uint8_t *b_p) {
+    uint8_t r, g, b, base, color;
+    r = g = b = 0; // To make the compiler stop complaining about potentially uninitialized variables
```
[the remark this answers is not in the document]
Comment (kwerdenker) [reply quantum/rgblight.c]: Has been fixed with 78178f7

File: quantum/rgblight.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (kwerdenker) [reply quantum/rgblight.c]: Has been addressed with c044c4b

File: quantum/rgblight.c [hunk 5]
```diff
@@ -57,79 +57,117 @@ rgblight_config_t rgblight_config;
...
-  r = pgm_read_byte(&CIE1931_CURVE[r]);
-  g = pgm_read_byte(&CIE1931_CURVE[g]);
-  b = pgm_read_byte(&CIE1931_CURVE[b]);
 
+    if (sat == 0) { // Acromatic color (gray). Hue doesn't mind.
+      r = val;
+      g = val;
+      b = val;
+    } else {
+      while(hue > 360) { hue -= 360; } //Making sure Hue is at or below 360
```
Comment (vomindoraan) [inline quantum/rgblight.c]: I believe this is just
```suggestion
      hue %= 360;
```

File: quantum/rgblight.c [hunk 6]
```diff
@@ -57,79 +57,117 @@ rgblight_config_t rgblight_config;
...
+
+      base = ((255 - sat) * val) >> 8;
```
Comment (vomindoraan) [inline quantum/rgblight.c]: Might also want to check that `sat < 256`.

File: quantum/rgblight.c [hunk 7]
```diff
@@ -57,79 +57,117 @@ rgblight_config_t rgblight_config;
...
+          b = val;
+          break;
+        case 5:
+          r = val;
+          g = base;
+          b = val - color;
+          break;
+        default:
+          //We should never be here but the compiler likes it thorough
+          r = 0;
```
Comment (vomindoraan) [inline quantum/rgblight.c]: I'd prefer `r = g = b = 0;` as you originally had, just for the sake of saving some vertical space, but feel free to ignore me.

File: quantum/rgblight.c [hunk 8]
```diff
@@ -57,79 +57,117 @@ rgblight_config_t rgblight_config;
...
-        g = base;
-        b = val;
-        break;
-      case 5:
-        r = val;
-        g = base;
-        b = val - color;
-        break;
+    if (val > RGBLIGHT_LIMIT_VAL) {
+        val=RGBLIGHT_LIMIT_VAL; // limit the val
```
Comment (vomindoraan) [inline quantum/rgblight.c]:
```suggestion
        val = RGBLIGHT_LIMIT_VAL; // limit the val
```

File: quantum/rgblight.c [hunk 9]
```diff
@@ -57,79 +57,117 @@ rgblight_config_t rgblight_config;
...
     }
-  }
-  r = pgm_read_byte(&CIE1931_CURVE[r]);
-  g = pgm_read_byte(&CIE1931_CURVE[g]);
-  b = pgm_read_byte(&CIE1931_CURVE[b]);
 
+    if (sat == 0) { // Acromatic color (gray). Hue doesn't mind.
+      r = val;
```
Comment (vomindoraan) [inline quantum/rgblight.c]: I'd prefer `r = g = b = val;` again just for brevity's sake.

File: quantum/rgblight.c (hunk 5 above)
Comment (kwerdenker) [reply quantum/rgblight.c]: You're right. That's a brain fart on my end.

File: quantum/rgblight.c (hunk 6 above)
Comment (kwerdenker) [reply quantum/rgblight.c]: `sat` is defined as uint8_t which makes it automatically `0 <= sat <= 255` or is there something I'm missing here?

File: quantum/rgblight.c (hunk 7 above)
Comment (kwerdenker) [reply quantum/rgblight.c]: Fine either way for me

File: quantum/rgblight.c (hunk 6 above)
Comment (vomindoraan) [reply quantum/rgblight.c]: Oh you're totally right, I missed that. Feel free to resolve.

**Comment by kwerdenker** [comment]
Fixed the remarks from the last code review. From my point of view the work for this PR is done. Any other work should be moved to a refactoring PR, since it doesn't involve the code that's added with the new feature and thus shouldn't block the merge.

**Comment by drashna** [comment]
Would you mind updating this?

**Comment by drashna** [comment]
For @skullydazed the #2350 PR also adds better/proper RGBW implementation. 

I think this implementation may be better, but the other one also splits up some stuff (bitbanging rgbw over twi)

**Comment by kwerdenker** [comment]
> 
> 
> For @skullydazed the #2350 PR also adds better/proper RGBW implementation.
> 
> I think this implementation may be better, but the other one also splits up some stuff (bitbanging rgbw over twi)

As far as I can tell the RGBW support in the other PR is limited to setting the W part always to 0 while this actually calculates the proper values to be used. I'm working on getting this PR up to date to the latest master branch version but I'm not too sure how to handle the merge of both PRs.
Maybe first merge my changes and then adapt the other PR to that? Seems easier that way from my perspective.

**Comment by kwerdenker** [comment]
I updated again after the latest changes on master but I can't keep doing this because someone changes something in rgblight.h/.c . So can we _please_ finally merge this?

**Comment by Taudris** [comment]
> As far as I can tell the RGBW support in the other PR is limited to setting the W part always to 0 while this actually calculates the proper values to be used.

I didn't include automatic calculation of W because it could introduce unwanted behavior for people. I implemented this for my personal layout, and one of the things I noticed on both of my ErgoDox EZ Shine (one original, one with CIY switches) is that at intermediate saturations, the left half flickers, even with a USB3.0 port. The cable just can't supply enough power. Given this, I didn't want to risk damaging people's keyboards, or even making them temporarily inoperable in case their systems had worse behavior than mine. My keyboard still operates fine when the LEDs are flickering, but this may not be true for someone using, say, a weak laptop USB port or an unpowered USB hub.

Even if this issue didn't happen, adding W to the mix changes the appearance of the output pretty significantly, potentially unpleasantly, and in ways that can't be solved purely algorithmically. Thus, I didn't want to make this decision for everyone. Always setting W to 0 at least fixes the buggy behavior when you do use it; for example, turning RGB off actually turns it off instead of turning RGB off and leaving W on.

Between the two problems above, I suggest you include a `#define` to enable automatic W so that people who don't want the change aren't forced to live with it. Remember that up until now, QMK users have lived without full W support and have been perfectly happy. Boards, layouts, and users that want it can easily enable it.

I also never actually use desaturation on my boards anyway; I paid for color, I want color! But that's just me. I only got W working because it was mildly interesting from a technical POV; the driver refactor was more important to me so I could work toward asymmetric RGB (which I gave up on when I thought my PR had been abandoned... perhaps I should give it a shot after all).

> I'm working on getting this PR up to date to the latest master branch version but I'm not too sure how to handle the merge of both PRs.

The only things my PR does in `rgblight.c` that yours doesn't are:

1. Change gamma to be global instead of only being used for effects (this would affect users' custom colors, but it cleans up the code quite a bit doing it globally, so I'm torn)
2. Use the unified LED API with `LED_TYPE` (no more separate `ws2812_setleds_rgbw`)
3. Call the new drivers separately:

```
#ifdef RGBW_BB_TWI
ledbbtwi_setleds(start_led, num_leds);
#endif

ws2812_setleds(start_led, num_leds);
```

Obviously, you can't do 2 or 3 until #2350 is merged.

Anyway, long story short, if this PR gets merged before mine, I'm happy to reapply my changes to the result. It will take me some time to resolve the comments on #2350.

**Comment by kwerdenker** [comment]
> I didn't include automatic calculation of W because it could introduce unwanted behavior for people. I implemented this for my personal layout, and one of the things I noticed on both of my ErgoDox EZ Shine (one original, one with CIY switches) is that at intermediate saturations, the left half flickers, even with a USB3.0 port. The cable just can't supply enough power. Given this, I didn't want to risk damaging people's keyboards, or even making them temporarily inoperable in case their systems had worse behavior than mine. My keyboard still operates fine when the LEDs are flickering, but this may not be true for someone using, say, a weak laptop USB port or an unpowered USB hub.

I've never had these problems in any of my tests. Since amount of white, that's set in the W portion of the struct, is taken out of all other colors the energy consumption of the whole LED shouldn't drastically change. In fact for colors with a large white part this change might even reduce the energy consumption because the white LED can produce the light more efficiently than by using the three color LEDs to mix it.

> Even if this issue didn't happen, adding W to the mix changes the appearance of the output pretty significantly, potentially unpleasantly, and in ways that can't be solved purely algorithmically. Thus, I didn't want to make this decision for everyone. Always setting W to 0 at least fixes the buggy behavior when you do use it; for example, turning RGB off actually turns it off instead of turning RGB off and leaving W on.

This change is the whole point of the exercise? RGBW LEDs are used to be able to produce better white colors than with just RGB LEDs and if the hardware supports it, it should be used by the software.

> Between the two problems above, I suggest you include a `#define` to enable automatic W so that people who don't want the change aren't forced to live with it. Remember that up until now, QMK users have lived without full W support and have been perfectly happy. Boards, layouts, and users that want it can easily enable it.

QMK already had a '#define RGBW'  before my change, it just didn't do anything besides changing the LED_TYPE to a four byte array. With this change enabling this actually makes use of the W part, which I'd say is what any user who sets that option expects anyways. (I know I did when first looking into this ;) )

> I also never actually use desaturation on my boards anyway; I paid for color, I want color! But that's just me. I only got W working because it was mildly interesting from a technical POV; the driver refactor was more important to me so I could work toward asymmetric RGB (which I gave up on when I thought my PR had been abandoned... perhaps I should give it a shot after all).

My intereste in this was purely because I used RGBW LEDs in a project and was suprised when things didn't work after setting (what I thought was) the appropriate '#define'. I've researched a lot of different ways of handling calculating the W part out of RGB colors to make this work with as little impact on existing functions as possible.

> Anyway, long story short, if this PR gets merged before mine, I'm happy to reapply my changes to the result. It will take me some time to resolve the comments on #2350.

 Thank you for offering this solution. Due to time constraints on my part I would obviously prefer this way around but if there are project reasons to do it the other way around, I'll try to make it work

**Comment by Taudris** [comment]
> I've never had these problems in any of my tests. Since amount of white, that's set in the W portion of the struct, is taken out of all other colors the energy consumption of the whole LED shouldn't drastically change. In fact for colors with a large white part this change might even reduce the energy consumption because the white LED can produce the light more efficiently than by using the three color LEDs to mix it.

My logic does the same channel subtraction that yours is doing. The flickering I've seen tells me that the W LEDs definitely take more power than the other colors do. It stands to reason, seeing as they are so much brighter than the RGB LEDs. I don't know why you aren't seeing the same behavior. All the more reason to be cautious.

> This change is the whole point of the exercise? RGBW LEDs are used to be able to produce better white colors than with just RGB LEDs and if the hardware supports it, it should be used by the software.

It produces a better white, sure, but everything in between is worse. The white LEDs are brighter than the RGB LEDs, so they cause excessive desaturation. Additionally, the white point of the white LEDs is different than what you get at 100% RGB, so there will be a color shift.

> QMK already had a '#define RGBW' before my change, it just didn't do anything besides changing the LED_TYPE to a four byte array. With this change enabling this actually makes use of the W part, which I'd say is what any user who sets that option expects anyways. (I know I did when first looking into this ;) )

The RGBW define tells QMK to send 32 bits for each LED instead of 24. LEDs that have RGBW require this regardless of one's desire to use W or not; hence why a new `#define` would be required to toggle automatic W.

**Comment by XScorpion2** [comment]
PR thoughts for consideration:
- Not a fan of the API changes, `sethsv `to `convert_hsv_to_rgb` doesn't gain you anything in the hot code paths at the cost of increasing code size. (pointer is 16 bits)
- The additional API's to spell out using RGBW vs RGB seem a bit excessive as using LED_TYPE these can just be made generic.
- I like the automatic W channel calculation, but the rest of the PR feels like it misses the mark a bit, adding new API's to do something that should just me agnostic.

**Comment by drashna** [comment]
Well, for what it's worth, I don't see flickering on mine. But that may be because of the power source (directly connected to a PC)

**Comment by drashna** [comment]
[Proper RGBW support for RGB Light feature](https://trello.com/c/FCmHVZvU/14-proper-rgbw-support-for-rgb-light-feature)

**Comment by drashna** [comment]
It looks like there are some merge conflicts now, due to a number of changes to the rgblight code.

**Comment by kwerdenker** [comment]
> It looks like there are some merge conflicts now, due to a number of changes to the rgblight code.

Based on review responses and the fact that a lot of things have changed around in the RGB code for QMK, I feel like the best solution at this point is to wipe the slate clean and start fresh.

What's your prefered way of doing this: closing this PR and opening a new one or keeping this open and force-pushing the rework into it?

**Comment by drashna** [comment]
I think closing and re-opening it would be the best way to go, yeah. 

Also, on the plus side(?), rgb and rgbw are a unified function

**Comment by drashna** [comment]
I've reworked this PR into a form that works with the newer code base. 
And I would like to git this merged in "soon", due to issues with the Ergodox EZ, specifically. 

The commit for it is: 
https://github.com/drashna/qmk_firmware/commit/6bf09d48f86c8200441402b56c30eb74ef275f93

I can open a new PR with this change, if you'd like.  And I can tag you as the co-author if you'd like.  Or you can use this commit.  Just let me know, and then we can move forward on this. 

Also, this addresses all of @XScorpion2's concerns, I think, as we've recently unified and cleaned up the ws2812 driver code.  So, it's one function for both rgb and rgbw, based on a define.   (as you can see)

**Comment by kwerdenker** [comment]
> The commit for it is:
> [drashna@6bf09d4](https://github.com/drashna/qmk_firmware/commit/6bf09d48f86c8200441402b56c30eb74ef275f93)
Made a comment on your commit

> I can open a new PR with this change, if you'd like. And I can tag you as the co-author if you'd like. Or you can use this commit. Just let me know, and then we can move forward on this.
> 
> Also, this addresses all of @XScorpion2's concerns, I think, as we've recently unified and cleaned up the ws2812 driver code. So, it's one function for both rgb and rgbw, based on a define. (as you can see)

You can handle it however it's best for you or the project, I don't mind

**Comment by drashna** [comment]
Sounds good.  My main concern was making sure that credit ends up where it is due.  With you.   You wrote the code, I just updated it to work with the updates that have been added.

**Comment by drashna** [comment]
Closing this for cleanup.  

And thank you for the initial contribution!!
